### PR TITLE
Dynamic Steps

### PR DESCRIPTION
**Pull request by justinmc**
This PR will be my first stab at generating the "tasks" steps seen numbered on the frontend dynamically, based on data passed from the server with the workflow.  I'll try to limit the backend impact of this for now.

Issues https://github.com/Autodesk/molecular-simulation-tools/issues/204 and https://github.com/Autodesk/molecular-simulation-tools/issues/205

~~Waiting on https://github.com/Autodesk/molecular-simulation-tools/pull/228 and https://github.com/Autodesk/molecular-simulation-tools/pull/229~~

### CODE REVIEW

**Comment by justinmc** [comment]
@PeterRJones Check out the migration setup I wrote for this: https://github.com/Autodesk/molecular-simulation-tools/pull/231/commits/31327b8fb4667564b2cb78e10f918ac1256ec02f#diff-7cd651498ecb2e6d55bc2ea9aa2785e0R30

How do you normally handle db migrations?  Is the production db nicely backed up in a way that we can roll back if these migrations destroy everything?  It worked on my local db but I'm nervous :)

File: client/public/js/components/tasks.jsx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+  return (
+    <div className="tasks-pane">
+      <div key={0} className="tasks">
+        <ol>
+          {
+            props.app.tasks.map((task, index) => {
+              const number = index + 1;
+
+              switch (task.id) {
```
Comment (justinmc) [inline client/public/js/components/tasks.jsx]: The tasks array is iterated, and task ids are converted into an actual UI button (left column) in a big switch statement.  Might be moved into a utils file later, but the general idea is the same.

screen shot 2017-03-22 at 10 17 52 am

File: client/public/js/utils/task_utils.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    });
+  },
+
+  /**
+   * Given a task (and run data), return a bool indicating if it's completed
+   * @param {TaskRecord}
+   * @param {RunRecord}
+   * @return {Boolean}
+   */
+  isCompleted(task, run) {
```
Comment (justinmc) [inline client/public/js/utils/task_utils.js]: Logic for determining the completion state of a task is here.  There is no separate state for marking that a task is completed.  It's just determined by looking at run state (i.e. for Load, if input pdb data exists, then Load must be done).

File: client/public/js/components/status.jsx [hunk 3]
```diff
@@ -115,32 +35,70 @@ function Status(props) {
...
-        <StatusLigandSelection
-          changeLigandSelection={props.changeLigandSelection}
-          ligandNames={ioUtils.getLigandNames(props.workflow.run.inputs)}
-          runCompleted={runCompleted}
-          selectedLigand={props.selectedLigand}
-        />
-      );
+    } else if (!props.app.fetching && !props.app.fetchingError &&
+      props.selection.type === selectionConstants.TASK) {
+      switch (props.app.tasks.get(props.selection.taskIndex).id) {
```
Comment (justinmc) [inline client/public/js/components/status.jsx]: The status pane finds its content the same way as the numbered steps in the left column.  A switch statement on the task id runs some custom code for each task id.

screen shot 2017-03-22 at 10 19 10 am

File: server/utils/seed_data.js [hunk 4]
```diff
@@ -10,6 +12,14 @@ const workflows = [
     creatorImage: '/img/logo2.png',
     description: 'Calculate the electron binding energy of an' +
         ' anionic doublet species using DFT',
+    tasks: [
```
Comment (justinmc) [inline server/utils/seed_data.js]: The only task data we're using for now is the task ID, so that's all I'm including in the app data.

**Comment by justinmc** [comment]
@dionjwa Please take a look at this as a first step on the way towards using the dynamic app data that we've been discussing.  I tried to add comments on the interesting parts since the PR is huge (due to renaming `workflow` to `app` everywhere).  I left the name `tasks` as is since we didn't seem conclusive on a final name yesterday.  Also notice that I'm purposely not yet using an intermediate VERIFY step for loading a PDB.

File: client/public/js/components/app_card.jsx [hunk 5]
```diff
@@ -72,17 +74,17 @@ function WorkflowCard(props) {
...
 
-WorkflowCard.propTypes = {
+AppCard.propTypes = {
   bgColor: React.PropTypes.string.isRequired,
   bgIndex: React.PropTypes.number.isRequired,
   color: React.PropTypes.string.isRequired,
   comingSoon: React.PropTypes.bool.isRequired,
-  creatorImage: React.PropTypes.string.isRequired,
+  // creatorImage: React.PropTypes.string.isRequired,
   id: React.PropTypes.string.isRequired,
```
Comment (dionjwa) [inline client/public/js/components/app_card.jsx]: Just remove?

**Comment by dionjwa** [comment]
@justinmc This is too big with too many changes, about changes that are probably only partially done, for me to review this effectively. I'm going to merge, but you'll need to guide me through these current changes and the proposed changes.

**Pull request merged**